File: fe/netlify/functions/predict.py
import json
import joblib
import numpy as np
import pandas as pd
import os

# Get the directory where this function is located
FUNCTION_DIR = os.path.dirname(os.path.abspath(__file__))

# Load the trained model and scaler from the function directory
model = None
scaler = None

def load_models():
    global model, scaler
    if model is None or scaler is None:
        model_path = os.path.join(FUNCTION_DIR, 'heart_disease_model.pkl')
        scaler_path = os.path.join(FUNCTION_DIR, 'scaler.pkl')
        model = joblib.load(model_path)
        scaler = joblib.load(scaler_path)
    return model, scaler

# Define the column names for the features (13 features expected)
COLUMNS = [
    'age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 
    'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal'
]

def handler(event, context):
    # Handle CORS preflight
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Content-Type': 'application/json'
    }
    
    # Handle OPTIONS request (CORS preflight)
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    # Only allow POST requests
    if event.get('httpMethod') != 'POST':
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    try:
        # Parse request body
        body = json.loads(event.get('body', '{}'))
        
        # Validate required fields
        missing_fields = [col for col in COLUMNS if col not in body]
        if missing_fields:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': f'Missing required fields: {", ".join(missing_fields)}'})
            }
        
        # Extract features
        features = [float(body[col]) for col in COLUMNS]
        
        # Load models
        model, scaler = load_models()
        
        # Convert input data to a pandas DataFrame
        input_data = pd.DataFrame([features], columns=COLUMNS)
        
        # Scale the input data
        input_data_scaled = scaler.transform(input_data)
        
        # Make prediction
        prediction = model.predict(input_data_scaled)[0]
        
        # Get prediction probability
        probability = model.predict_proba(input_data_scaled)[0]
        
        # Format the result
        result = "Heart Disease Detected" if prediction == 1 else "No Heart Disease"
        confidence = f"{max(probability) * 100:.2f}"
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'prediction': result,
                'confidence': confidence
            })
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': str(e)})
        }

File: fe/netlify/functions/predict.js
const { Handler } = require('@netlify/functions');

// Pre-computed coefficients from the trained model
// These would need to be extracted from your scikit-learn model
const MODEL_COEFFICIENTS = {
  intercept: -5.5,
  weights: {
    age: 0.03,
    sex: 1.2,
    cp: 0.8,
    trestbps: 0.02,
    chol: 0.005,
    fbs: 0.5,
    restecg: 0.3,
    thalach: -0.03,
    exang: 1.0,
    oldpeak: 0.5,
    slope: 0.4,
    ca: 0.9,
    thal: 0.7
  }
};

// Scaler parameters (mean and std for each feature)
const SCALER_PARAMS = {
  mean: {
    age: 54.4, sex: 0.68, cp: 0.97, trestbps: 131.6, chol: 246.3,
    fbs: 0.15, restecg: 0.53, thalach: 149.6, exang: 0.33,
    oldpeak: 1.04, slope: 1.4, ca: 0.73, thal: 2.31
  },
  std: {
    age: 9.0, sex: 0.47, cp: 1.03, trestbps: 17.5, chol: 51.8,
    fbs: 0.36, restecg: 0.53, thalach: 22.9, exang: 0.47,
    oldpeak: 1.16, slope: 0.62, ca: 1.02, thal: 0.61
  }
};

function sigmoid(x) {
  return 1 / (1 + Math.exp(-x));
}

function scaleFeature(value, feature) {
  return (value - SCALER_PARAMS.mean[feature]) / SCALER_PARAMS.std[feature];
}

function predict(features) {
  // Scale features
  const scaledFeatures = {};
  for (const [key, value] of Object.entries(features)) {
    scaledFeatures[key] = scaleFeature(value, key);
  }
  
  // Calculate linear combination
  let linearCombination = MODEL_COEFFICIENTS.intercept;
  for (const [feature, weight] of Object.entries(MODEL_COEFFICIENTS.weights)) {
    linearCombination += scaledFeatures[feature] * weight;
  }
  
  // Apply sigmoid to get probability
  const probability = sigmoid(linearCombination);
  
  // Make prediction (threshold at 0.5)
  const prediction = probability >= 0.5 ? 1 : 0;
  
  return {
    prediction: prediction === 1 ? "Heart Disease Detected" : "No Heart Disease",
    confidence: (Math.max(probability, 1 - probability) * 100).toFixed(2),
    probability: probability.toFixed(4)
  };
}

exports.handler = async (event, context) => {
  // CORS headers
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Content-Type': 'application/json'
  };

  // Handle preflight
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  // Only allow POST
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      headers,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    
    // Required fields
    const requiredFields = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal'];
    
    // Check for missing fields
    const missingFields = requiredFields.filter(field => !(field in body));
    if (missingFields.length > 0) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: `Missing fields: ${missingFields.join(', ')}` })
      };
    }

    // Extract features
    const features = {};
    for (const field of requiredFields) {
      features[field] = parseFloat(body[field]);
      if (isNaN(features[field])) {
        return {
          statusCode: 400,
          headers,
          body: JSON.stringify({ error: `Invalid value for ${field}` })
        };
      }
    }

    // Make prediction
    const result = predict(features);

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify(result)
    };

  } catch (error) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: error.message })
    };
  }
};
